readme.md
# pdfsnag

Streaming htmltopdf web service

![pdfsnag](https://raw.github.com/applieddataconsultants/pdfsnag/master/pdfsnag.png)

## Setup

Requires that you have [wkhtmltopdf](http://code.google.com/p/wkhtmltopdf) binary in your PATH.  See below for instructions on how to set that up on Linux server environment.

To start server run:

```
sudo npm install -g pdfsnag
pdfsnag [port]
```

### Installing wkhtmltopdf standalone (w/out X)

```sh
sudo aptitude install openssl build-essential xorg libssl-dev # ubuntu specific
```

For amd64

```sh
wget http://wkhtmltopdf.googlecode.com/files/wkhtmltopdf-0.9.9-static-amd64.tar.bz2
tar xvjf wkhtmltopdf-0.9.9-static-amd64.tar.bz2
mv wkhtmltopdf-amd64 /usr/local/bin/wkhtmltopdf
chmod +x /usr/local/bin/wkhtmltopdf
```

For i386

```sh
wget http://wkhtmltopdf.googlecode.com/files/wkhtmltopdf-0.9.9-static-i386.tar.bz2
tar xvjf wkhtmltopdf-0.9.9-static-i386.tar.bz2
mv wkhtmltopdf-i386 /usr/local/bin/wkhtmltopdf
chmod +x /usr/local/bin/wkhtmltopdf
```

## License

(The MIT License)

Copyright (c) 2012 Applied Data Consultants

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish, distribute,
sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


changelog.md
1.0.1 / 2016-07-28
==================

-	Fix linux stdout issue

1.0.0 / 2016-07-28
==================

-	Add eslintrc and remove Makefile
-	Update support for newer versions of wkhtmltopdf and Node

0.1.5 / 2012-12-10
==================

-	more aggressively stopping wkhtmltopdf on error
-	added; image

0.1.4 / 2012-11-02
==================

-	added; logging
-	minor; updated makefile

0.1.3 / 2012-08-27
==================

-	fixed; url encode/decode issue
-	added; forkme

0.1.1 / 2012-04-16
==================

-	added; allow for mixture of get and post values

0.1.0 / 2012-04-11
==================

-	added; ability to pass in custom options


app.js
const spawn = require('child_process').spawn
const http = require('http')
const url = require('url')
const qs = require('querystring')
const port = process.argv[2] || 8003

function Undefined(){}
function StringStripped(value){ return value.replace('$','').replace('&','').replace(';','') }

const acceptableOpts = {
   'background': Undefined,
   'collate': Undefined,
   'copies': Number,
   'default-header': Undefined,
   'disable-external-links': Undefined,
   'disable-forms': Undefined,
   'disable-internal-links': Undefined,
   'disable-javascript': Undefined,
   'disable-smart-shrinking': Undefined,
   'disable-toc-back-links': Undefined,
   'enable-external-links': Undefined,
   'enable-forms': Undefined,
   'enable-internal-links': Undefined,
   'enable-javascript': Undefined,
   'enable-smart-shrinking': Undefined,
   'enable-toc-back-links': Undefined,
   'encoding': StringStripped,
   'exclude-from-outline': Undefined,
   'grayscale': Undefined,
   'image-dpi': Number,
   'image-quality': Number,
   'images': Undefined,
   'include-in-outline': Undefined,
   'javascript-delay': Number,
   'load-error-handling': StringStripped,
   'lowquality': Undefined,
   'margin-bottom': StringStripped,
   'margin-left': StringStripped,
   'margin-right': StringStripped,
   'margin-top': StringStripped,
   'minimum-font-size': Number,
   'no-background': Undefined,
   'no-collate': Undefined,
   'no-images': Undefined,
   'no-outline': Undefined,
   'no-pdf-compression': Undefined,
   'no-print-media-type': Undefined,
   'no-stop-slow-scripts': Undefined,
   'orientation': StringStripped,
   'outline': Undefined,
   'outline-depth': Number,
   'output-format': StringStripped,
   'page-height': StringStripped,
   'page-offset': Number,
   'page-size': StringStripped,
   'page-width': StringStripped,
   'password': StringStripped,
   'print-media-type': Undefined,
   'proxy': StringStripped,
   'stop-slow-scripts': Undefined,
   'title': StringStripped,
   'user-style-sheet': StringStripped,
   'username': StringStripped,
   'zoom': Number
}

const index = require('fs').readFileSync(__dirname+'/index.html').toString().replace('%OPTIONS%', Object.keys(acceptableOpts))

function afterwards (res, wkhtmltopdf) {
   if (!res.finished) {
      wkhtmltopdf.kill('SIGTERM')
      res.end('Unable to load within 30 seconds')
   }
}

function getopts (query) {
   var opts = []
   Object.keys(query).forEach(function (opt) {
      var Type = acceptableOpts[opt]
      if (Type) {
         opts.push('--'+opt)
         if (Type != Undefined) opts.push(Type(query[opt]))
      }
   })

   return opts
}

function snagit (query, res) {
   query.html || (query.html = null)
   query.url  || (query.url  = null)
   query.name || (query.name = 'output')

   var opts = getopts(query)

   if (!query.html && !query.url) {
      res.writeHead(200, { 'Content-Type': 'text/html' })
      res.end(index)
   }
   else {
      res.writeHead(200, {
         'Content-Type': 'application/pdf',
         'Content-Disposition': 'attachment; filename='+query.name+'.pdf'
      })

      var wkhtmltopdf = spawn('wkhtmltopdf',[ (query.html ? '-' : query.url),'-'].concat(opts))
      if (query.html) wkhtmltopdf.stdin.end(query.html)
      wkhtmltopdf.stdout.pipe(res)
      setTimeout(afterwards, 30000, res, wkhtmltopdf)
   }
}

http.createServer( function (req, res) {
   var query = url.parse(req.url, true).query

   if (req.method === 'POST') {
      req.setEncoding('utf-8')
      var body = ''
      req.on('data', function (chunk) {
         body += chunk
      })
      req.on('end', function () {
         query = qs.parse(body)
         snagit(query, res)
      })
   } else snagit(query, res)

}).listen(port)


pdfsnag.js
#!/usr/bin/env node

var spawn = require('child_process').spawn
var http = require('http')
var url = require('url')
var qs = require('querystring')
var port = process.argv[2] || 3000
var util = require('util')

function Undefined() {}
function StringStripped(value) { return value.replace('$','').replace('&','').replace(';','') }

var acceptableOpts = {
  'background': Undefined,
  'collate': Undefined,
  'copies': Number,
  'default-header': Undefined,
  'disable-external-links': Undefined,
  'disable-forms': Undefined,
  'disable-internal-links': Undefined,
  'disable-javascript': Undefined,
  'disable-smart-shrinking': Undefined,
  'disable-toc-back-links': Undefined,
  'enable-external-links': Undefined,
  'enable-forms': Undefined,
  'enable-internal-links': Undefined,
  'enable-javascript': Undefined,
  'enable-smart-shrinking': Undefined,
  'enable-toc-back-links': Undefined,
  'encoding': StringStripped,
  'exclude-from-outline': Undefined,
  'grayscale': Undefined,
  'image-dpi': Number,
  'image-quality': Number,
  'images': Undefined,
  'include-in-outline': Undefined,
  'javascript-delay': Number,
  'load-error-handling': StringStripped,
  'lowquality': Undefined,
  'margin-bottom': StringStripped,
  'margin-left': StringStripped,
  'margin-right': StringStripped,
  'margin-top': StringStripped,
  'minimum-font-size': Number,
  'no-background': Undefined,
  'no-collate': Undefined,
  'no-images': Undefined,
  'no-outline': Undefined,
  'no-pdf-compression': Undefined,
  'no-print-media-type': Undefined,
  'no-stop-slow-scripts': Undefined,
  'orientation': StringStripped,
  'outline': Undefined,
  'outline-depth': Number,
  'output-format': StringStripped,
  'page-height': StringStripped,
  'page-offset': Number,
  'page-size': StringStripped,
  'page-width': StringStripped,
  'password': StringStripped,
  'print-media-type': Undefined,
  'proxy': StringStripped,
  'stop-slow-scripts': Undefined,
  'title': StringStripped,
  'user-style-sheet': StringStripped,
  'username': StringStripped,
  'zoom': Number,
}

var index = require('fs').readFileSync(__dirname + '/index.html').toString().replace('%OPTIONS%', Object.keys(acceptableOpts))
var forkme = require('fs').readFileSync(__dirname + '/forkme.png')

function afterwards(res, wkhtmltopdf) {
  if (!res.finished) {
    util.log('error - wkhtmltopdf did not complete in a timely manner, killing')
    wkhtmltopdf.stdout.destroy()
    wkhtmltopdf.stderr.destroy()
    wkhtmltopdf.kill('SIGKILL')
  }
}

function getopts(query) {
  var opts = []
  Object.keys(query).forEach(function(opt) {
    var Type = acceptableOpts[opt]
    if (Type) {
      opts.push('--' + opt)
      if (Type != Undefined) opts.push(Type(query[opt]))
    }
  })

  return opts
}

function snagit(query, res) {
  query.html || (query.html = null)
  query.url || (query.url = null)
  query.name || (query.name = 'output')

  var opts = getopts(query)

  res.writeHead(200, {
    'Content-Type': 'application/pdf',
    'Content-Disposition': 'attachment; filename=' + query.name + '.pdf',
  })

  var wkhtmltopdf = spawn('/bin/sh', ['-c', 'wkhtmltopdf ' + opts.join(' ') + ' ' + (query.html ? '-' : '"' + encodeURI(decodeURI(query.url)) + '"') + ' - | cat'])
  if (query.html) wkhtmltopdf.stdin.end(query.html)
  wkhtmltopdf.stdout.pipe(res)
  util.log('info - ' + query.name + ' ' + (query.url || '"' + query.html.substr(25, 50).replace(/\n|\r/g,'') + '"') + ' ' + JSON.stringify(opts))
  setTimeout(afterwards, 30000, res, wkhtmltopdf)
}

http.createServer(function(req, res) {
  var query = url.parse(req.url, true).query

  if (req.method === 'POST') {
    req.setEncoding('utf-8')
    var body = ''
    req.on('data', function(chunk) {
      body += chunk
    })
    req.on('end', function() {
      body = qs.parse(body)
      for (var i in body) query[i] = body[i]
      snagit(query, res)
    })
  } else if (!query.html && !query.url) {
    if (req.url == '/forkme.png') {
      res.writeHead(200, {'Content-Type': 'image/png'})
      res.end(forkme)
    }
    else {
      res.writeHead(200, {'Content-Type': 'text/html'})
      res.end(index)
    }
  }
  else snagit(query, res)

}).listen(port)
